planner/core/fragment.go

```diff
@@ -115,20 +134,107 @@ func (f *Fragment) init(p PhysicalPlan) error {
 	return nil
 }
 
-func newFragment(s *PhysicalExchangeSender) (*Fragment, error) {
-	f := &Fragment{ExchangeSender: s}
-	s.Fragment = f
-	err := f.init(s)
-	return f, errors.Trace(err)
+// We would remove all the union-all operators by 'untwist'ing and copying the plans above union-all.
+// This will make every route from root (ExchangeSender) to leaf nodes (ExchangeReceiver and TableScan)
+// a new ioslated tree (and also a fragment) without union all. These trees (fragments then tasks) will
+// finally be gathered to TiDB or be exchanged to upper tasks again. 
+// For instance, given a plan "select c1 from t union all select c1 from s"
+// after untwist, there will be two plans in `forest` slice:
+// - ExchangeSender -> Projection (c1) -> TableScan(t)
+// - ExchangeSender -> Projection (c2) -> TableScan(s)
+func untwistPlanAndRemoveUnionAll(stack []PhysicalPlan, forest *[]*PhysicalExchangeSender) error {
+	cur := stack[len(stack)-1]
+	switch x := cur.(type) {
+	case *PhysicalTableScan, *PhysicalExchangeReceiver: // This should be the leave node.
+		p, err := stack[0].Clone()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		*forest = append(*forest, p.(*PhysicalExchangeSender))
+		for i := 1; i < len(stack); i++ {
+			if _, ok := stack[i].(*PhysicalUnionAll); ok {
+				continue
+			}
+			ch, err := stack[i].Clone()
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if join, ok := p.(*PhysicalHashJoin); ok {
+				join.SetChild(1-join.InnerChildIdx, ch)
+			} else {
+				p.SetChildren(ch)
+			}
+			p = ch
+		}
+	case *PhysicalHashJoin:
+		stack = append(stack, x.children[1-x.InnerChildIdx])
+		err := untwistPlanAndRemoveUnionAll(stack, forest)
+		stack = stack[:len(stack)-1]
+		return errors.Trace(err)
+	case *PhysicalUnionAll:
+		for _, ch := range x.children {
+			stack = append(stack, ch)
+			err := untwistPlanAndRemoveUnionAll(stack, forest)
+			stack = stack[:len(stack)-1]
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+	default:
+		if len(cur.Children()) != 1 {
+			return errors.New("unexpected plan " + cur.ExplainID().String())
```

Comment:
```suggestion
			return errors.Trace(errors.New("unexpected plan " + cur.ExplainID().String()))
```
And replace above `return errors.Trace(err)` to `return err`.